Explore deprecation of istanbul-api

I'd like for us to consider deprecating `istanbul-api` in favor of consumers directly using the implementation modules (`istanbul-lib-report`, etc).  The motivation is `source-map@0.7.3` - the async constructor requires breaking API changes to `istanbul-lib-source-maps` and I'm not sure how to make `istanbul-api` work with it.  One idea I have is to publish `istanbul-api@3.0.0` and immediate deprecate that version only.  This way we discourage new modules from using istanbul-api but we don't cause a deprecation warning for the existing major users.  When we drop node.js 6 and upgrade `source-map` we would not update `istanbul-api` to be compatible.

According to npmjs.com `istanbul-api` was downloaded 5.1 million times last week.  `jest-cli` was downloaded 4.5 million times.  `jest-cli` is being split up so the `istanbul-api` dependency is moving to `@jest/reporters`.  `createReporter` is the only thing jest uses from `istanbul-api` so I'm hoping we can convince them to use `istanbul-lib-report` directly.

Another large user is `karma-coverage-istanbul-reporter` at 700k downloads last week.  `fusion-cli` had 15k last week, I'm not seeing any other dependents with a large number of downloads.

About the missing 200k `istanbul-api` downloads, I'm guessing `jest-cli@14` and below is still getting significant downloads, that version uses the original `istanbul` module.

@SimenB I think I have the correct code in place for jest-reporters, having an issue with typescript checking.
```
packages/jest-reporters/src/coverage_reporter.ts:102:43 - error TS2345: Argument of type 'string' is not assignable to parameter of type '"none" | "text-summary" | "clover" | "cobertura" | "html" | "json" | "json-summary" | "lcov" | "lcovonly" | "teamcity" | "text" | "text-lcov"'.

102         tree.visit(istanbulReports.create(_reporter, {}), reportContext);
                                              ~~~~~~~~~

packages/jest-cli/src/cli/index.ts:14:22 - error TS2307: Cannot find module '@jest/core'.
```

It seems that `@types/istanbul-report` declares the specific values which can be used.  The broken branch is at https://github.com/coreyfarrell/jest/tree/reports-next.  Note that I have no idea what to do about `packages/jest-reporters/src/__tests__/coverage_reporter.test.js`.  For now I've just deleted that file.

--- a/packages/istanbul-api/README.md
+++ b/packages/istanbul-api/README.md
@@ -3,4 +3,12 @@
 [![Greenkeeper badge](https://badges.greenkeeper.io/istanbuljs/istanbul-api.svg)](https://greenkeeper.io/)
 [![Build Status](https://travis-ci.org/istanbuljs/istanbul-api.svg?branch=master)](https://travis-ci.org/istanbuljs/istanbul-api)
 
-High-level API for istanbul.
+Deprecated high-level API for istanbul.
+
+Remaining users of this module should migrate:
+
+-   istanbul-lib-instrument or babel-plugin-istanbul - modify sources to include coverage counters
+-   istanbul-lib-coverage - merge coverage results from multiple tests into a single dataset for reporting
+-   istanbul-lib-source-maps - process source-maps before generating reports
+-   istanbul-lib-report - the reporting system
+-   istanbul-reports - the actual reports (text-summary report, html report, etc)

